src/session/content/message_row/media.rs

```diff
@@ -0,0 +1,203 @@
+
+mod imp {
+    use super::*;
+    use once_cell::sync::Lazy;
+
+    #[derive(Debug, Default, CompositeTemplate)]
+    #[template(resource = "/com/github/melix99/telegrand/ui/content-message-media.ui")]
+    pub struct Media {
+        #[template_child]
+        pub content: TemplateChild<gtk::Box>,
+        #[template_child]
+        pub picture: TemplateChild<MediaPicture>,
+        #[template_child]
+        pub caption_label: TemplateChild<gtk::Label>,
+        #[template_child]
+        pub progress_bar: TemplateChild<gtk::ProgressBar>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for Media {
+        const NAME: &'static str = "ContentMessageMedia";
+        type Type = super::Media;
+        type ParentType = gtk::Widget;
+
+        fn class_init(klass: &mut Self::Class) {
+            Self::bind_template(klass);
+        }
+
+        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
+            obj.init_template();
+        }
+    }
+
+    impl ObjectImpl for Media {
+        fn properties() -> &'static [glib::ParamSpec] {
+            static PROPERTIES: Lazy<Vec<glib::ParamSpec>> = Lazy::new(|| {
+                vec![
+                    glib::ParamSpec::new_string(
+                        "caption",
+                        "Caption",
+                        "The caption",
+                        None,
+                        glib::ParamFlags::READWRITE | glib::ParamFlags::EXPLICIT_NOTIFY,
+                    ),
+                    glib::ParamSpec::new_double(
+                        "download-progress",
+                        "Download Progress",
+                        "The download progress",
+                        0.0,
+                        1.0,
+                        0.0,
+                        glib::ParamFlags::READWRITE | glib::ParamFlags::EXPLICIT_NOTIFY,
+                    ),
+                ]
+            });
+            PROPERTIES.as_ref()
+        }
+
+        fn set_property(
+            &self,
+            obj: &Self::Type,
+            _id: usize,
+            value: &glib::Value,
+            pspec: &glib::ParamSpec,
+        ) {
+            match pspec.name() {
+                "caption" => obj.set_caption(value.get().unwrap()),
+                "download-progress" => obj.set_download_progress(value.get().unwrap()),
+                _ => unimplemented!(),
+            }
+        }
+
+        fn property(&self, obj: &Self::Type, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
+            match pspec.name() {
+                "caption" => obj.caption().to_value(),
+                "download-progress" => obj.download_progress().to_value(),
+                _ => unimplemented!(),
+            }
+        }
+
+        fn constructed(&self, obj: &Self::Type) {
+            self.parent_constructed(obj);
+
+            self.progress_bar.connect_fraction_notify(|bar| {
+                bar.set_visible(bar.fraction() < 1.0);
+            });
```

Comment:
As I understand, the signal handler will still be there after the media has been downloaded. But it has no purpose any longer. I think it's better to put this in `Media::set_download_progress` instead.
In this we, we would not need a handler.